index.rst

```diff
@@ -51,11 +51,182 @@
+It is also possible that a special remote Datastore subclass could be written that reimplements the few methods that require direct table access (``getStoredItemsInfo``, ``addStoredItemsInfo``, ``removeStoredItemsInfo``, and ``_registered_refs_per_artifact``) as REST calls on the server.
+
+The dataset artifact side of Datastore already knows how to transfer data to and from a remote object store or file server via the generic ``ButlerURI`` implementation so there is no need to duplicate that support into a special butler server.
+
+What will be required though is to include in the butler server a means for translating the URIs stored in the datastore registry into `signed URLs`_.
+The URIs would likely be written as ``gcs://`` or ``s3://`` form in the registry and would be converted to signed ``https`` URLs for use by the client.
+For read this could be handled directly in the server implementation of ``getStoredItemsInfo`` such that the URIs are always returned as signed URLs.
+For write, the URI is constructed using the file template and ``LocationFactory`` called from ``Datastore._prepare_for_put()``.
+This code would have to be factored out into its own method such that a client ``Datastore`` subclass could ask the server to convert that URI to a signed URL.
+
+It has to be decided whether file templates are handled solely by the server configuration or if the client is allowed to specify local template modifications.
+
+
+.. _signed URLs: https://cloud.google.com/storage/docs/access-control/signed-urls
+
+Example Configuration
+^^^^^^^^^^^^^^^^^^^^^
+
+The Butler should be configured by the URL of the server and the server should have a ``butler.yaml`` file available at its root.
+This configuration file is not the configuration of the server itself but is the configuration that clients should use.
+Its contents should be simple and could be something like:
+
+.. code-block:: yaml
+
+  datastore:
+    cls: lsst.daf.butler.datastores.client.ClientDatastore
+    root: <butlerRoot>
+  registry:
+    cls: lsst.daf.butler.registry.RegistryClient
+    db: <butlerRoot>
+  storageClasses:
+    # Storage classes known to the server
+
+where ``<butlerRoot>`` would automatically be replaced by the URL of the server.
+
+Problems
+^^^^^^^^
+
+Transaction handling might be an issue since it is very hard to implement rollbacks of registry changes if there is a problem on the datastore side without requiring that the connection to the server stays open.
+This is a particular issue during ``Butler.put()`` and to a lesser extent ``Butler.pruneDatasets()``.
+This can be mitigated if datastore always does the put first before registry updates the tables.
+There may need to be a special API added to the registry server to allow all registry updates to happen once.
+The complication is that this level of change would require a new ``Butler`` implementation specifically for the user unless it was managed by a configuration option.
```

Comment:
Hmm.  It sure does sound like we'll need either a client/server `Butler` implementation as well as a client/server `Registry` and client/server `Datastore` - or a pretty significant rework of the boundary between `Registry` and `Datastore`.

That rework may be in order anyway given that one of the main ways we'll be adding to the `Datastore` (i.e. batch processing) looks like it'll be uploading files to object stores long before the associated database entries for them are committed.  And while I think we've convinced ourselves that the boundary we have can work with that order of operations, I think it's still a sign that the as-defined boundary isn't really a natural one anymore.  I have no organized thoughts on this, but my disorganized ones include:
 - Dropping `dataset_location`: if most datasets _usually_ exist in some datastore, and we can efficiently ask Datastores "which of these datasets do you actually have", then not being able to join against `dataset_location` in SQL isn't a big problem.
 - Using some scalable NoSQL database for datastore records.
 - Using UUIDs instead of autoincrement integers to relate Registry and Datastore (mostly so a stable, unique ID for a dataset can be created by a Datastore instead of Registry if the dataset lands there first).

What all of those (and this butler transaction problem) have in common is embracing the idea that a dataset can exist in just `Datastore`, just `Registry`, or both, so `Datastore` and `Registry` are both their own source of truth for dataset existence.

One aspect of this that I have not thought about at all is whether it helps with the signed URL stuff.

Reply:
I think from butler's point of view in the batch processing world, the datastore.put will be doing something entirely predictable using a dataset_id given to it from the entry already in registry.

I'm open to a situation where datastore disentangles itself from butler registry.

We mostly use dataset_location table for trash handling.